File: MD.CMS.BusinessLogic.Core/DataAccess/Entities/GenericContent/BaseField.cs
using MD.Tools.BaseDataAccess.Core.Entities;
using System;
using System.Xml;
using MD.CMS.BusinessLogic.Core.DataAccess.Entities.Field;

namespace MD.CMS.BusinessLogic.Core.DataAccess.Entities.GenericContent
{
    public abstract class BaseField : BaseEntity<long>
    {
        #region Attributes
        private long _attributeTypeDefinitionId;
        private string _name;
        private string _defaultValue;
        private XmlDocument _validationXml;
        private AttributeTypeDefinition _attributeTypeDefinition;
        private string _delimiter;
        private string _listValue;
        private string _options;
        private Options _jsonField;
        private ValidationType _validation;
        private string _uniqueId;
        #endregion

        #region Properties

        public long AttributeTypeDefinitionId
        {
            get { return _attributeTypeDefinitionId; }
            set { _attributeTypeDefinitionId = value; }
        }

        public string Name
        {
            get { return _name; }
            set { _name = value; }
        }

        public string FriendlyName
        {
            get
            {
                if (_name != null)
                {
                    return _name.Replace(" ", "_")
                      .Replace("$", "")
                      .Replace(".", "")
                      .Replace("@", "")
                      .Replace("{", "")
                      .Replace("}", "")
                      .Replace("[", "")
                      .Replace("]", "")
                      .Replace("\\", "")
                      .Replace("/", "")
                      .Replace("*", "")
                      .Replace("(", "")
                      .Replace(")", "")
                      .Replace("%", "")
                      .Replace("#", "")
                      .Replace("!", "")
                      .Replace("&", "")
                      .Replace("?", "")
                      .Replace("+", "")
                      .Replace("-", "")
                      .Replace("=", "");
                }
                return _name;

            }
        }

        public bool IsNew
        {
            get
            {
                return Id <= default(long);
            }
        }

        public AttributeTypeDefinition AttributeTypeDefinition
        {
            get
            {
                return _attributeTypeDefinition;
            }
            set
            {
                _attributeTypeDefinition = value;
                if (value != null && !value.Id.Equals(default(long)))
                {
                    _attributeTypeDefinitionId = value.Id;
                }
            }
        }

        public string DefaultValue
        {
            get { return _defaultValue; }
            set { _defaultValue = value; }
        }

        public string Options
        {
            get { return _options; }
            set { _options = value; }
        }

        public Options JsonField
        {
            get
            {
                if (_jsonField == null)
                {
                    Options options = new Options();
                    if (!string.IsNullOrEmpty(_options))
                    {
                        try
                        {
                            options = Newtonsoft.Json.JsonConvert.DeserializeObject<Options>(_options);
                        }
                        catch (Exception e)
                        {

                        }
                    }
                    if (options.validation == null)
                    {
                        options.validation = new Field.Validation.ValidationType();
                    }
                    _jsonField = options;
                }
                return _jsonField;
            }
            set
            {
                if (value != null)
                {
                    _options = Newtonsoft.Json.JsonConvert.SerializeObject(value);
                }
            }
        }

        public string Delimiter
        {
            get { return _delimiter; }
            set { _delimiter = value; }
        }

        public string ListValue
        {
            get { return _listValue; }
            set { _listValue = value; }
        }

        public virtual bool IsReadOnly
        {
            get
            {
                bool returnValue = false;

                if(JsonField != null && !JsonField.enabled)
                {
                    returnValue = true;
                }

                return returnValue;
            }
        }

        public virtual string UniqueId
        {
            get
            {
                if (string.IsNullOrEmpty(_uniqueId))
                {
                    _uniqueId = string.Format("_{0}", this.Id.ToString());
                }
                return _uniqueId;
            }
        }
        #endregion

        #region Methods
        public void Serialize()
        {
            _options = Newtonsoft.Json.JsonConvert.SerializeObject(JsonField);
        }
        #endregion

        #region Methods
        public BaseField() : base()
        {

        }

        public BaseField(BaseField obj) : base(obj)
        {
            if (obj != null)
            {
                _attributeTypeDefinitionId = obj.AttributeTypeDefinitionId;
                _name = obj.Name;
                _defaultValue = obj.DefaultValue;
                _attributeTypeDefinition = obj.AttributeTypeDefinition;
                _delimiter = obj.Delimiter;
                _listValue = obj.ListValue;
                _options = obj.Options;
                _jsonField = obj.JsonField;
                _validation = obj._validation;
                _uniqueId = obj.UniqueId;
            }
        }
        #endregion
    }
}

File: MD.CMS.Administration/MD.CMS.Administration.Core.Web/scripts/businessLogic/dataAccess/entities/genericContent/baseField.ts
/// <reference path="../attributeTypeDefinition.ts" />
namespace mdBusinessLogic.dataAccess.entities.genericContent {
    export abstract class baseField extends base.BaseEntity {
        public AttributeTypeDefinitionId: number;
        public Name: string;
        public DefaultValue: string;
        public AttributeTypeDefinition: attributeTypeDefinition;
        public Delimiter: string;
        public ListValue: string;
        public FriendlyName: string;
        public IsRequired: boolean;
        public UniqueId: string;
        public IsReadOnly: boolean;

        constructor(obj?: baseField) {
            super(obj);
            this.AttributeTypeDefinitionId = 0;
            this.AttributeTypeDefinition = null;
            this.Name = '';
            this.IsRequired = false;
            this.DefaultValue = '';
            this.Delimiter = '';
            this.ListValue = '';
            this.FriendlyName = '';
            this.UniqueId = '';
            this.IsReadOnly = false;

            if (obj != undefined && obj != null) {
                this.construct(obj);
            } else {
                if (this.ListValue == '[]') {
                    this.ListValue = '';
                }
                if (this.DefaultValue == '[]') {
                    this.DefaultValue = '';
                }

                if ((this.DefaultValue === undefined || this.DefaultValue == null || this.DefaultValue == '') && !(this.ListValue === undefined || this.ListValue == null || this.ListValue == '')) {
                    this.DefaultValue = this.getListValueAsArray()[0] || '';
                }
            }
        }

        public construct(data: any) {
            super.construct(data);
            this.AttributeTypeDefinitionId = this.getValue<number>(data, 'AttributeTypeDefinitionId', 0);
            this.AttributeTypeDefinition = this.getConstructEntityValue<attributeTypeDefinition>(data, 'AttributeTypeDefinition', new attributeTypeDefinition());
            this.Name = this.getValue<string>(data, 'Name', '');
            this.IsRequired = this.getValue<boolean>(data, 'IsRequired', false);
            this.DefaultValue = this.getValue<string>(data, 'DefaultValue', '');
            this.FriendlyName = this.getValue<string>(data, 'FriendlyName', '');
            this.UniqueId = this.getValue<string>(data, 'UniqueId', '');
            this.Delimiter = this.getValue<string>(data, 'Delimiter', '');
            this.ListValue = this.getValue<string>(data, 'ListValue', '');
            this.IsReadOnly = this.getValue<boolean>(data, 'IsReadOnly', false);

            if (this.ListValue == '[]') {
                this.ListValue = '';
            }
            if (this.DefaultValue == '[]') {
                this.DefaultValue = '';
            }

            if ((this.DefaultValue === undefined || this.DefaultValue == null || this.DefaultValue == '') && !(this.ListValue === undefined || this.ListValue == null || this.ListValue == '')) {
                this.DefaultValue = this.getListValueAsArray()[0] || '';
            }
        }

        public getListValueAsArray(): Array<string> {
            if (this.ListValue === undefined || this.ListValue == null) {
                this.ListValue = '';
            }
            if (this.Delimiter === undefined || this.Delimiter == null) {
                this.Delimiter = '';
            }
            return this.ListValue.split(this.Delimiter);
        }
    }
}
